From: EdgePi E87N firmware build <noreply@example.invalid>
Date: Wed, 5 Aug 2026 00:00:00 +0800
Subject: [PATCH] staging: fbtft: add NV3007 display controller

Add the NV3007 panel controller used by the EdgePi E87N.  The panel
initialization, address offsets, rotation mapping and gamma programming
match the working vendor runtime.
---
 drivers/staging/fbtft/Kconfig     |   7 ++
 drivers/staging/fbtft/Makefile    |   1 +
 drivers/staging/fbtft/fb_nv3007.c | 217 ++++++++++++++++++++++++++++++
 3 files changed, 225 insertions(+)
 create mode 100644 drivers/staging/fbtft/fb_nv3007.c

diff --git a/drivers/staging/fbtft/Kconfig b/drivers/staging/fbtft/Kconfig
--- a/drivers/staging/fbtft/Kconfig
+++ b/drivers/staging/fbtft/Kconfig
@@ -77,6 +77,13 @@ config FB_TFT_ILI9486
 	help
 	  Generic Framebuffer support for ILI9486
 
+config FB_TFT_NV3007
+	tristate "FB driver for the NV3007 LCD Controller"
+	depends on FB_TFT
+	help
+	  Generic framebuffer support for the New Vision NV3007 LCD
+	  controller.
+
 config FB_TFT_PCD8544
 	tristate "FB driver for the PCD8544 LCD Controller"
 	depends on FB_TFT
diff --git a/drivers/staging/fbtft/Makefile b/drivers/staging/fbtft/Makefile
--- a/drivers/staging/fbtft/Makefile
+++ b/drivers/staging/fbtft/Makefile
@@ -17,6 +17,7 @@ obj-$(CONFIG_FB_TFT_ILI9340)     += fb_ili9340.o
 obj-$(CONFIG_FB_TFT_ILI9341)     += fb_ili9341.o
 obj-$(CONFIG_FB_TFT_ILI9481)     += fb_ili9481.o
 obj-$(CONFIG_FB_TFT_ILI9486)     += fb_ili9486.o
+obj-$(CONFIG_FB_TFT_NV3007)      += fb_nv3007.o
 obj-$(CONFIG_FB_TFT_PCD8544)     += fb_pcd8544.o
 obj-$(CONFIG_FB_TFT_RA8875)      += fb_ra8875.o
 obj-$(CONFIG_FB_TFT_S6D02A1)     += fb_s6d02a1.o
diff --git a/drivers/staging/fbtft/fb_nv3007.c b/drivers/staging/fbtft/fb_nv3007.c
new file mode 100644
--- /dev/null
+++ b/drivers/staging/fbtft/fb_nv3007.c
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * FB driver for the New Vision NV3007 LCD Controller
+ */
+
+#include <linux/delay.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+
+#include "fbtft.h"
+
+#define DRVNAME "fb_nv3007"
+#define WIDTH 428
+#define HEIGHT 142
+
+#define DEFAULT_GAMMA \
+	"00 02 04 07 05 02 21 23 3A 08 13 13 29 31 0F\n" \
+	"00 02 04 07 05 02 21 23 3A 08 13 13 29 31 0F"
+
+static int init_display(struct fbtft_par *par)
+{
+	fbtft_par_dbg(DEBUG_INIT_DISPLAY, par, "%s()\n", __func__);
+
+	par->fbtftops.reset(par);
+
+	write_reg(par, 0x11);
+	mdelay(5);
+	write_reg(par, 0x28);
+	write_reg(par, 0xFF, 0xA5);
+	write_reg(par, 0x9A, 0x08);
+	write_reg(par, 0x9B, 0x08);
+	write_reg(par, 0x9C, 0xB0);
+	write_reg(par, 0x9D, 0x17);
+	write_reg(par, 0x9E, 0xC2);
+	write_reg(par, 0x8F, 0x22, 0x04);
+	write_reg(par, 0x84, 0x90);
+	write_reg(par, 0x83, 0x7B);
+	write_reg(par, 0x85, 0x4F);
+	write_reg(par, 0x50, 0x00);
+	write_reg(par, 0x52, 0xD6);
+	write_reg(par, 0x53, 0x04);
+	write_reg(par, 0x54, 0x04);
+	write_reg(par, 0x55, 0x1B);
+	write_reg(par, 0x56, 0x1B);
+	write_reg(par, 0xA0, 0x2A, 0x24, 0x00);
+	write_reg(par, 0xA1, 0x84);
+	write_reg(par, 0xA2, 0x85);
+	write_reg(par, 0xA8, 0x34);
+	write_reg(par, 0xA9, 0x80);
+	write_reg(par, 0xAA, 0x73);
+	write_reg(par, 0xAB, 0x03, 0x61);
+	write_reg(par, 0xAC, 0x03, 0x65);
+	write_reg(par, 0xAD, 0x03, 0x60);
+	write_reg(par, 0xAE, 0x03, 0x64);
+	write_reg(par, 0xB9, 0x82);
+	write_reg(par, 0xBA, 0x83);
+	write_reg(par, 0xBB, 0x80);
+	write_reg(par, 0xBC, 0x81);
+	write_reg(par, 0xBD, 0x02);
+	write_reg(par, 0xBE, 0x01);
+	write_reg(par, 0xBF, 0x04);
+	write_reg(par, 0xC0, 0x03);
+	write_reg(par, 0xC4, 0x33);
+	write_reg(par, 0xC5, 0x80);
+	write_reg(par, 0xC6, 0x73);
+	write_reg(par, 0xC7, 0x00);
+	write_reg(par, 0xC8, 0x33, 0x33);
+	write_reg(par, 0xC9, 0x5B);
+	write_reg(par, 0xCA, 0x5A);
+	write_reg(par, 0xCB, 0x5D);
+	write_reg(par, 0xCC, 0x5C);
+	write_reg(par, 0xCD, 0x33, 0x33);
+	write_reg(par, 0xCE, 0x5F);
+	write_reg(par, 0xCF, 0x5E);
+	write_reg(par, 0xD0, 0x61);
+	write_reg(par, 0xD1, 0x60);
+	write_reg(par, 0xB0, 0x3A, 0x3A, 0x00, 0x00);
+	write_reg(par, 0xB6, 0x32);
+	write_reg(par, 0xB7, 0x80);
+	write_reg(par, 0xB8, 0x73);
+	write_reg(par, 0xE0, 0x00);
+	write_reg(par, 0xE1, 0x03, 0x0F);
+	write_reg(par, 0xE2, 0x04);
+	write_reg(par, 0xE3, 0x01);
+	write_reg(par, 0xE4, 0x0E);
+	write_reg(par, 0xE5, 0x01);
+	write_reg(par, 0xE6, 0x19);
+	write_reg(par, 0xE7, 0x10);
+	write_reg(par, 0xE8, 0x10);
+	write_reg(par, 0xE9, 0x21);
+	write_reg(par, 0xEA, 0x12);
+	write_reg(par, 0xEB, 0xD0);
+	write_reg(par, 0xEC, 0x04);
+	write_reg(par, 0xED, 0x07);
+	write_reg(par, 0xEE, 0x07);
+	write_reg(par, 0xEF, 0x09);
+	write_reg(par, 0xF0, 0xD0);
+	write_reg(par, 0xF1, 0x0E);
+	write_reg(par, 0xF9, 0x56);
+	write_reg(par, 0xF2, 0x26, 0x1B, 0x0B, 0x20);
+	write_reg(par, 0xEC, 0x04);
+	write_reg(par, 0x35, 0x00);
+	write_reg(par, 0x44, 0x00, 0x10);
+	write_reg(par, 0x46, 0x10);
+	write_reg(par, 0xFF, 0x00);
+	write_reg(par, 0x3A, 0x05);
+	write_reg(par, 0x11);
+	mdelay(200);
+	write_reg(par, 0x29);
+	mdelay(150);
+	write_reg(par, 0x2A, 0x00, 0x0C, 0x00, 0x99);
+	write_reg(par, 0x2B, 0x00, 0x00, 0x01, 0xAB);
+	mdelay(20);
+
+	return 0;
+}
+
+static void set_addr_win(struct fbtft_par *par,
+			 int xs, int ys, int xe, int ye)
+{
+	switch (par->info->var.rotate) {
+	case 0:
+		xs += 14;
+		xe += 14;
+		break;
+	case 90:
+		ys += 14;
+		ye += 14;
+		break;
+	case 180:
+		xs += 12;
+		xe += 12;
+		break;
+	case 270:
+		ys += 12;
+		ye += 12;
+		break;
+	default:
+		break;
+	}
+
+	write_reg(par, 0x2A, xs >> 8, xs & 0xFF, xe >> 8, xe & 0xFF);
+	write_reg(par, 0x2B, ys >> 8, ys & 0xFF, ye >> 8, ye & 0xFF);
+	write_reg(par, 0x2C);
+}
+
+static int set_var(struct fbtft_par *par)
+{
+	switch (par->info->var.rotate) {
+	case 0:
+		write_reg(par, 0x36, 0xC0 | (par->bgr << 3));
+		break;
+	case 90:
+		write_reg(par, 0x36, 0x60 | (par->bgr << 3));
+		break;
+	case 180:
+		write_reg(par, 0x36, 0x00 | (par->bgr << 3));
+		break;
+	case 270:
+		write_reg(par, 0x36, 0xA0 | (par->bgr << 3));
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static int set_gamma(struct fbtft_par *par, u32 *curves)
+{
+	unsigned int curve;
+	unsigned int value;
+
+	write_reg(par, 0xFF, 0xA5);
+	for (curve = 0; curve < par->gamma.num_curves; curve++) {
+		for (value = 0; value < par->gamma.num_values; value++) {
+			write_reg(par, 0x60 + curve * 0x10 + value,
+				  curves[curve * par->gamma.num_values + value]);
+		}
+	}
+	write_reg(par, 0xFF, 0x00);
+
+	return 0;
+}
+
+static int blank(struct fbtft_par *par, bool on)
+{
+	fbtft_par_dbg(DEBUG_BLANK, par, "%s(blank=%d)\n", __func__, on);
+	write_reg(par, on ? 0x28 : 0x29);
+
+	return 0;
+}
+
+static struct fbtft_display display = {
+	.width = 428,
+	.height = 142,
+	.regwidth = 8,
+	.buswidth = 8,
+	.bpp = 16,
+	.gamma = DEFAULT_GAMMA,
+	.gamma_num = 2,
+	.gamma_len = 15,
+	.fbtftops = {
+		.init_display = init_display,
+		.set_addr_win = set_addr_win,
+		.set_var = set_var,
+		.set_gamma = set_gamma,
+		.blank = blank,
+	},
+};
+
+FBTFT_REGISTER_DRIVER(DRVNAME, "newvisionu,nv3007", &display);
+
+MODULE_ALIAS("spi:" DRVNAME);
+MODULE_ALIAS("platform:" DRVNAME);
+MODULE_DESCRIPTION("FB driver for the New Vision NV3007 LCD Controller");
+MODULE_LICENSE("GPL");
-- 
2.50.1
